### DISCUSSION TITLE
"Array device must be same as the current device" issue "with nlp.pipe()"

### DISCUSSION THREAD

**Discussion by fatihbozdag** [explosion/spaCy#11455] [Help: Coding & Implementations] [answered]
Greetings,

I am running Spacy on Colab. However, I got the error` 'Array device must be same as the current device: array device = -2 while current = 0' when using` `nlp.pipe()`

code as follows:

```
!pip install -U pip setuptools wheel
!pip install -U spacy
!pip install spacy_transformers -U
import spacy
import spacy_transformers
spacy.require_gpu()
!python -m spacy download('en_core_web_trf')
nlp = spacy.load('en_core_web_trf')

TIME = []
TIME_LEMMA = []
TIME_START = []
TIME_NATIVE_LANG = []

for doc, context in nlp_pipe(CORPUS as_tuples =True):
    for token in doc:
        if token.dep_ == 'advmod' and token.lemma_ in TIME_SUBORDINATORS:
            TIME.append(token.sent)
            TIME_LEMMA.append(token.lemma_)
            TIME_START.append(token.is_sent_start)
            TIME_NATIVE_LANG.append(context['Native_language'])

cupy/_core/core.pyx in cupy._core.core.ndarray.copy()
cupy/_core/core.pyx in cupy._core.core.ndarray.copy()
cupy/_core/core.pyx in cupy._core.core.ndarray.astype()
cupy/_core/_kernel.pyx in cupy._core._kernel.ufunc.__call__()
cupy/_core/_kernel.pyx in cupy._core._kernel._preprocess_args()
cupy/_core/_kernel.pyx in cupy._core._kernel._check_array_device_id()
ValueError: Array device must be same as the current device: array device = -2 while current = 0
```
However, without `for doc, context in nlp_pipe(CORPUS, as_tuples =True):` code simply works.

```
    for token in doc:
        if token.dep_ == 'advmod' and token.lemma_ in TIME_SUBORDINATORS:
            TIME.append(token.sent)
            TIME_LEMMA.append(token.lemma_)
            TIME_START.append(token.is_sent_start)
            TIME_NATIVE_LANG.append(context['Native_language'])
```
So, I assume I should define GPU on `nlp.pipe()` ? I haven't had this issue on my local computer, though. Also without importing `spacy_transformers`, `I got no GPU detected`

**Comment by fatihbozdag** [comment, marked as answer]
Solved! by replacing code from
```
spacy.require_gpu()
nlp = spacy.load('en_core_web_trf')
```
to
```
nlp = spacy.load('en_core_web_trf')
spacy.require_gpu()
```

I am not sure about the logic here and do not know if it was mentioned before, but _accidentally_ changing the lines worked!

**Reply by polm** [reply]
Glad you figured it out, thanks for sharing the solution! I am also not sure why that would work but it may have to do with some of the special handling for GPU support colab has - we'll look into it!

**Reply by fatihbozdag** [reply]
Sure thing! I've added some details [here](https://github.com/explosion/spaCy/discussions/11330#discussioncomment-3593544) for another user having the same problems. I guess it is more to do with Jupyter than Colab GPU handling.
